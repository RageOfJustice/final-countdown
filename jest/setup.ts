import 'jest-styled-components'

jest.mock('react-native/Libraries/YellowBox/YellowBox')
