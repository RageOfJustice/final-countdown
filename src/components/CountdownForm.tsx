import React, { useState, useRef } from 'react'
import { TextInput, Keyboard } from 'react-native'
import styled from 'src/styled-components'

export type CountdownFormMode = 'start' | 'stop'
export type CountdownFormPressHandler = (minutes?: number) => void

interface Props {
  mode: CountdownFormMode
  onPress?: CountdownFormPressHandler
}

const CountdownForm: React.FC<Props> = ({ mode, onPress }) => {
  const [value, setValue] = useState('')

  const inputRef = useRef<TextInput>(null)

  const isInStartState = mode === 'start'

  const handlePress = () => {
    if (!isInStartState) {
      onPress?.()
    }
    const parsed = parseInt(value, 10)
    if (!isNaN(parsed)) {
      onPress?.(parsed)
    }
    setValue('')
    inputRef.current?.clear()
    Keyboard.dismiss()
  }

  return (
    <Wrapper>
      <Text>Countdown:</Text>
      <Input
        ref={inputRef}
        value={value}
        editable={isInStartState}
        onChangeText={setValue}
        placeholder="Minutes"
        keyboardType="decimal-pad"
      />
      <ButtonWrapper
        disabled={isInStartState && !value}
        onPress={handlePress}
        testID="controlButton"
      >
        <ButtonText>{isInStartState ? 'Start' : ' Stop'}</ButtonText>
      </ButtonWrapper>
    </Wrapper>
  )
}

const Wrapper = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: center;
`

const Text = styled.Text`
  color: ${({ theme }) => theme.colors.black};
`

const Input = styled.TextInput.attrs(({ theme }) => ({
  underlineColorAndroid: 'transparent',
  placeholderTextColor: theme.colors.lightGrey,
  selectionColor: theme.colors.darkGrey,
}))`
  color: ${({ theme }) => theme.colors.black};
  margin-horizontal: 8px;
  border-width: 1px;
  padding: 8px 4px;
  width: 100px;
  border-color: ${({ theme }) => theme.colors.black};
  opacity: ${({ editable }) => (editable ? 1 : 0.6)};
`

const ButtonWrapper = styled.TouchableOpacity`
  padding: 8px;
  opacity: ${({ disabled }) => (disabled ? 0.6 : 1)};
  background-color: ${({ theme }) => theme.colors.cyan};
`

const ButtonText = styled.Text`
  color: ${({ theme }) => theme.colors.white};
`

export default CountdownForm
