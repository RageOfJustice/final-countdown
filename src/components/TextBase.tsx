import { Platform } from 'react-native'

import styled from 'src/styled-components'

// на некоторых андроидах обрезаются слова
// поэтому надо явно указать шрифт
const TextBase = styled.Text`
  ${Platform.OS === 'android' && 'font-family: Roboto;'}
`

export default TextBase
