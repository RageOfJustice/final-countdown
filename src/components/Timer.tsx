import React, {
  memo,
  useRef,
  useEffect,
  forwardRef,
  useReducer,
  useCallback,
  useImperativeHandle,
} from 'react'
import styled from 'src/styled-components'

// TODO: вынести в helpers/formatTime и протестировать
const formatTime = (seconds: number): string => {
  if (seconds === 0) {
    return '00:00'
  }
  const minutes = Math.trunc(seconds / 60)
  const restSeconds = Math.trunc(seconds % 60)

  const formattedMinutes = `${minutes < 10 ? '0' : ''}${minutes}`
  const formattedSeconds = `${restSeconds < 10 ? '0' : ''}${restSeconds}`

  return `${formattedMinutes}:${formattedSeconds}`
}

interface Props {
  initialSeconds?: number
  speedMultiplier?: number
  onTimeUp?: () => void
  onHalfTimePassed?: () => void
}

export interface TimerInstance {
  start: () => void
  pause: () => void
  stop: () => void
}

type State = number
type Action =
  | { type: 'reset' }
  | { type: 'decrease' }
  | { type: 'setup'; seconds: number }

function reducer(state: State, action: Action): State {
  switch (action.type) {
    case 'reset':
      return 0
    case 'setup':
      return action.seconds
    case 'decrease':
      return state - 1
  }
}

const Timer = forwardRef<TimerInstance, Props>(
  (
    { onTimeUp, onHalfTimePassed, speedMultiplier = 1, initialSeconds = 0 },
    ref,
  ) => {
    const intervalId = useRef<number>()
    const halfTimeCalled = useRef(false)
    const [secondsLeft, dispatch] = useReducer(reducer, initialSeconds)

    const startTimer = useCallback(() => {
      if (!intervalId.current) {
        let interval
        if (speedMultiplier === 0) {
          interval = 1000
        } else {
          interval = Math.trunc(1000 / speedMultiplier)
        }
        const updateTime = () => {
          dispatch({ type: 'decrease' })
        }

        intervalId.current = setInterval(updateTime, interval)
      }
    }, [speedMultiplier, initialSeconds])

    const pauseTimer = useCallback(() => {
      clearInterval(intervalId.current!)
      intervalId.current = undefined
    }, [])

    const stopTimer = useCallback(() => {
      pauseTimer()
      halfTimeCalled.current = false
      dispatch({ type: 'reset' })
    }, [pauseTimer])

    useImperativeHandle(
      ref,
      () => ({
        start: startTimer,
        pause: pauseTimer,
        stop: stopTimer,
      }),
      [startTimer, pauseTimer, stopTimer],
    )

    useEffect(() => {
      const shouldContinue = !!intervalId.current
      if (shouldContinue) {
        pauseTimer()
      }

      dispatch({ type: 'setup', seconds: initialSeconds })

      if (shouldContinue) {
        startTimer()
      }
    }, [initialSeconds])

    useEffect(() => {
      if (intervalId.current) {
        pauseTimer()
        startTimer()
      }
    }, [speedMultiplier])

    useEffect(() => {
      if (secondsLeft <= 0 && intervalId.current) {
        stopTimer()
        onTimeUp?.()
        return
      }
      if (
        !halfTimeCalled.current &&
        secondsLeft < Math.ceil(initialSeconds / 2)
      ) {
        halfTimeCalled.current = true
        onHalfTimePassed?.()
      }
    }, [secondsLeft, onTimeUp, onHalfTimePassed, initialSeconds])

    return <TimeText>{formatTime(secondsLeft)}</TimeText>
  },
)

const TimeText = styled.Text`
  font-size: 80px;
  text-align: center;
  color: ${({ theme }) => theme.colors.black};
`

export default memo(Timer)
