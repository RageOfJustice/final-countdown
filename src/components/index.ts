export { default as Toggles } from './Toggles'
export { default as CountdownForm } from './CountdownForm'
