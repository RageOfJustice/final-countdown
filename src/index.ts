import React from 'react'

interface Props {}

const App: React.FC<Props> = props => {
  return null
}

export default App
