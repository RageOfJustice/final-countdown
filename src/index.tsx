import React from 'react'
import { ThemeProvider } from './styled-components'
import theme from './theme'

interface Props {}

const App: React.FC<Props> = () => {
  return <ThemeProvider theme={theme} />
}

export default App
