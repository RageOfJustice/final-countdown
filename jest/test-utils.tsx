import React, { ReactElement, FC } from 'react'
import { render, RenderOptions } from '@testing-library/react-native'
import theme from 'src/theme'
import { ThemeProvider } from 'src/styled-components'

const Providers: FC = ({ children }) => {
  return <ThemeProvider theme={theme}>{children as ReactElement}</ThemeProvider>
}

const customRender = (
  ui: ReactElement<any>,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react-native'

export { customRender as render }
