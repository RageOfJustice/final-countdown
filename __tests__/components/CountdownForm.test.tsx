import React from 'react'
import { render, fireEvent } from '../../jest/test-utils'
import { CountdownForm } from 'src/components'

describe('CountdownForm', () => {
  it('renders correctly in start mode', () => {
    const { container } = render(<CountdownForm mode="start" />)

    expect(container.children[0]).toMatchSnapshot()
  })

  it('renders correctly in start stop', () => {
    const { container } = render(<CountdownForm mode="stop" />)

    expect(container.children[0]).toMatchSnapshot()
  })

  it('behaves right in start mode', () => {
    const onPress = jest.fn()
    const { getByTestId, getByDisplayValue } = render(
      <CountdownForm onPress={onPress} mode="start" />,
    )

    const input = getByDisplayValue('')
    const button = getByTestId('controlButton')

    fireEvent.changeText(input, 'test')
    fireEvent.press(button)

    expect(onPress).not.toHaveBeenCalled()

    fireEvent.changeText(input, '12')

    fireEvent.press(button)
    expect(onPress).toHaveBeenCalledWith(12)
  })

  it('behaves right in stop mode', () => {
    const onPress = jest.fn()
    const { getByTestId } = render(
      <CountdownForm onPress={onPress} mode="stop" />,
    )

    const button = getByTestId('controlButton')

    fireEvent.press(button)
    expect(onPress).toHaveBeenCalledWith()
  })
})
