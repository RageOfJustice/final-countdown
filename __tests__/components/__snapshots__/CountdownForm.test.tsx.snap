// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountdownForm renders correctly in start mode 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#2d3436",
        },
      ]
    }
  >
    Countdown:
  </Text>
  <TextInput
    allowFontScaling={true}
    editable={true}
    keyboardType="decimal-pad"
    placeholder="Minutes"
    placeholderTextColor="#b2bec3"
    rejectResponderTermination={true}
    selectionColor="#636e72"
    style={
      Array [
        Object {
          "borderColor": "#2d3436",
          "borderWidth": 1,
          "color": "#2d3436",
          "marginHorizontal": 8,
          "opacity": 1,
          "paddingBottom": 8,
          "paddingLeft": 4,
          "paddingRight": 4,
          "paddingTop": 8,
          "width": 100,
        },
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <TouchableOpacity
    activeOpacity={0.2}
    disabled={true}
    style={
      Array [
        Object {
          "backgroundColor": "#00cec9",
          "opacity": 0.6,
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        },
      ]
    }
    testID="controlButton"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFFFFF",
          },
        ]
      }
    >
      Start
    </Text>
  </TouchableOpacity>
</View>
`;

exports[`CountdownForm renders correctly in start stop 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#2d3436",
        },
      ]
    }
  >
    Countdown:
  </Text>
  <TextInput
    allowFontScaling={true}
    editable={false}
    keyboardType="decimal-pad"
    placeholder="Minutes"
    placeholderTextColor="#b2bec3"
    rejectResponderTermination={true}
    selectionColor="#636e72"
    style={
      Array [
        Object {
          "borderColor": "#2d3436",
          "borderWidth": 1,
          "color": "#2d3436",
          "marginHorizontal": 8,
          "opacity": 0.6,
          "paddingBottom": 8,
          "paddingLeft": 4,
          "paddingRight": 4,
          "paddingTop": 8,
          "width": 100,
        },
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <TouchableOpacity
    activeOpacity={0.2}
    disabled={false}
    style={
      Array [
        Object {
          "backgroundColor": "#00cec9",
          "opacity": 1,
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        },
      ]
    }
    testID="controlButton"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFFFFF",
          },
        ]
      }
    >
       Stop
    </Text>
  </TouchableOpacity>
</View>
`;
